### PR TITLE
Developer-friendly ``pymc3.distributions`` (take 2)

### PR DESCRIPTION
**Pull request by superbobry**
This is #903 extended to continuous distributions.

Please do not merge just yet.

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -910,14 +935,14 @@ def __init__(self, distribution, lower=-inf, upper=inf):
...
+        return Bounded(first, self.distribution, self.lower, self.upper,
+                       *args, **kwargs)
 
     def dist(self, *args, **kwargs):
-        return Bounded.dist(self.distribution, self.lower, self.upper, *args, **kwargs)
+        return Bounded.dist(self.distribution, self.lower, self.upper,
+                            *args, **kwargs)
 
 
-Tpos = Bound(T, 0)
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: We'll probably want to keep that for convenience, just rename it to `StudentTpos`.

Comment (superbobry) [inline pymc3/distributions/continuous.py]: Done.

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -611,48 +634,47 @@ class Cauchy(Continuous):
...
+                                size=size)
 
     def logp(self, value):
         alpha = self.alpha
         beta = self.beta
-        return bound(
-            -log(pi) - log(beta) - log(1 + ((value - alpha) / beta) ** 2),
-            beta > 0)
+        return bound(- T.log(np.pi) - T.log(beta)
+                     - T.log(1 + ((value - alpha) / beta)**2),
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: T.log1p?

**Comment by superbobry** [comment]
All distributions accept `*args`, `**kwargs` in `__init__`. It seems that the only valid positional argument is `name` and the only valid keyword argument is `observed`. Am I right?

**Comment by twiecki** [comment]
@superbobry also `shape` and `transform`.

**Comment by twiecki** [comment]
@superbobry This is shaping up nicely. Let me know when you think it's ready for review.

**Comment by twiecki** [comment]
@superbobry ping. Have been looking at this and it seems fine to me. Any further changes you want us to wait on? Eager to get this merged before we deviate from it.

**Comment by superbobry** [comment]
I think this part is good to merge. The only module I haven't touched upon is `pymc3.distributions.distribution`, but I can do this in a separate PR.

**Comment by twiecki** [comment]
@fonnesbeck I think this looks great. Want to have a look?

**Comment by fonnesbeck** [comment]
This is all very helpful. I like it.

**Comment by twiecki** [comment]
The only wrinkle is that we're breaking backwards compat in an odd way. `T` is still around but not a distribution anymore. I wonder if we couldn't instead import `import theano.tensor as t` or `tensor` and make `T` raise a deprecation warning. Then after a little while we switch to `tensor as T`. Thoughts?

**Comment by superbobry** [comment]
I like the plan. However, it's not clear how long we should wait before we are safe to replace the `T`, since `pymc3` doesn't have proper releases atm.

**Comment by fonnesbeck** [comment]
I would definitely switch before a release. I would rate it as a pretty minor breaking change for a beta.

**Comment by twiecki** [comment]
Maybe it is best after all to just do it once rather than have it drag out. We changed all examples and could create an issue that people who look for it will be informed.

**Comment by twiecki** [comment]
@superbobry I went ahead and merged this as is. Thanks for the contribution, this really cleans up the code base a lot.

**Comment by twiecki** [comment]
Uploaded the new manual here: http://pymc-devs.github.io/pymc3/manual/api.html#distributions

**Pull request merged**